planning/behavior_velocity_planner/include/utilization/util.hpp

```diff
@@ -124,6 +125,29 @@ geometry_msgs::msg::Pose transformRelCoordinate2D(
   const geometry_msgs::msg::Pose & target, const geometry_msgs::msg::Pose & origin);
 geometry_msgs::msg::Pose transformAbsCoordinate2D(
   const geometry_msgs::msg::Pose & relative, const geometry_msgs::msg::Pose & origin);
+/**
+ * @brief calcSignedArcLength from point with closest index to point with with closest
+ */
+template <class T>
+double calcSignedArcLength(
+  const T & points, const size_t src_idx, const geometry_msgs::msg::Point & src_point,
+  const size_t dst_idx, const geometry_msgs::msg::Point & dst_point)
```

Comment:
My assumption is that it will be an argument to a function like this


```
struct PointWithSearchRange{
geometry_msgs::msg::Point point:
size_t min_path_index;
size_t max_path_index;
};


double calcSignedArcLengthFromFirstCloseSrcPointToDstPoint(
  const T & points, const size_t src_idx, const geometry_msgs::msg::Point & src_point,
  const PointWithSearchRange& dst_point, const float loop_judgment_distance=5)
```

Reply:
@yukkysaito 
I think each scene module have stop pose and stop pose closest index before calculating arc length but do you still want to recalculate closest index inside this function too?

Reply:
@taikitanaka3 cc @takayuki5168 
Or another sample code
```
struct PointWithSearchRange{
  geometry_msgs::msg::Point point:
  size_t min_path_index;
  size_t max_path_index;
};

PointWithSearchRange calcFirstClosePoint(
  const T & points, 
  const geometry_msgs::msg::Point & src_point, 
  const float loop_judgment_distance=5)

double calcSignedArcLength(
  const T & points, 
  const PointWithSearchRange & src_point, 
  const PointWithSearchRange& dst_point)

int main () {
  const auto & ego = src_point;
  PointWithSearchRange dst_point{stop_point, lane_id_min_path_index, lane_id_max_path_index};
  calcSignedArcLength(path.points, calcFirstClosePoint(path.points, ego), dst_point);
}

```

Reply:
@yukkysaito -san
first closest index and target closest index search is necessary so I've added these functions.
calcFirstClosePoint = findNearestIndex
if calcFirstClosePoint looks better I change this to it.
```
struct SearchIndex
{
  size_t first_idx;
  size_t last_idx;
};
struct PointWithSearchIndex
{
  geometry_msgs::msg::Point point;
  SearchIndex index;
};
SearchIndex getPathIndexRangeWithLaneId(
  const autoware_auto_planning_msgs::msg::PathWithLaneId & path,
  const geometry_msgs::msg::Point & point, const int64_t lane_id,
  const double distance_threshold = 3.0);
template <class T>
size_t findNearestIndex(const T & points, const PointWithSearchIndex & point_with_index)
template <class T>
size_t findNearestIndex(
  const T & points, const geometry_msgs::msg::Point & point, const double max_length = 10.0)
/**
 * @brief calcSignedArcLength from point to dst with search range
 */
template <class T>
double calcSignedArcLengthWithSearchIndex(
  const T & points, const geometry_msgs::msg::Point & src_point,
  const PointWithSearchIndex & dst_point_with_range)

```

Reply:
search range index

Reply:
memo
```
struct PointWithSearchRange{
  geometry_msgs::msg::Point point:
  size_t min_path_index;
  size_t max_path_index;
};

size_t calcFirstClosePointIndex(
  const T & points, 
  const geometry_msgs::msg::Point & src_point, 
  const float loop_judgment_distance=5)

double calcSignedArcLength(
  const T & points, 
  const size_t & src_index, 
  const PointWithSearchRange& dst_point)

int main () {
  const auto & ego = src_point;
  PointWithSearchRange dst_point{stop_point, lane_id_min_path_index, lane_id_max_path_index};
  const auto signed_arc_length = calcSignedArcLength(path.points, calcFirstClosePoint(path.points, ego), dst_point);
}




```

Reply:
memo
```
struct PointWithSearchRange{
  geometry_msgs::msg::Point point:
  size_t min_path_index;
  size_t max_path_index;
};

PointWithSearchRange calcFirstClosePoint(
  const T & points, 
  const geometry_msgs::msg::Point & src_point, 
  const float loop_judgment_distance=5)

size_t findNearestSegmentIndex(const T & points, const PointWithSearchRange & point) {
  // consider search range 
}


template <class T>
double calcSignedArcLength(
  const T & points, 
  const PointWithSearchRange t & src_point,
  const PointWithSearchRange & dst_point)
{
  validateNonEmpty(points);

  const size_t src_seg_idx = findNearestSegmentIndex(points, src_point);
  const size_t dst_seg_idx = findNearestSegmentIndex(points, dst_point);

  const double signed_length_on_traj = calcSignedArcLength(points, src_seg_idx, dst_seg_idx);
  const double signed_length_src_offset =
    calcLongitudinalOffsetToSegment(points, src_seg_idx, src_point);
  const double signed_length_dst_offset =
    calcLongitudinalOffsetToSegment(points, dst_seg_idx, dst_point);

  return signed_length_on_traj - signed_length_src_offset + signed_length_dst_offset;
}
```

Reply:
@yukkysaito 
updated at 
[a277bb0](https://github.com/autowarefoundation/autoware.universe/pull/405/commits/a277bb00f0f5660914520f8b56253644764ca31b)